### DISCUSSION TITLE
VideoPlayer

### DISCUSSION THREAD

**Discussion by firemen01** [DelphiWorlds/Kastri#256] [Q&A]
Please tell me if the videoplayer supports z-ordering?
a very useful thing

**Comment by DelphiWorlds** [comment]
The video player is based on the "native" implementation of FMX controls, so it supports z-ordering within other "native" controls

**Comment by firemen01** [comment]
I can’t figure out right away how I can display, for example, an image over a video?

```
  FPlayer := TVideoPlayer.Create;
  FPlayer.UseController:= false;
  FPlayer.View.Parent := myForm;
  FPlayer.View.Visible := true;
  FPlayer.Play(AURL);

  Image:= TImage.Create(myForm);
  Image.Parent:= myForm; 
  Image.Align:= TAlignLayout.Client;
  Image.Bitmap.LoadFromFile(myOverlayPNGFile);
```

video over image (

the same video player from Alсione based on the old echoplayer outputs a video texture with support for z-order and the picture is drawn on top, how to implement this correctly here?

**Comment by DelphiWorlds** [comment]
> the same video player from Alсione based on the old echoplayer outputs a video texture with support for z-order...

That's because Alcinoe uses a control based on TRectangle to display the video (which results in poorer performance), hence you're able to use regular FMX controls on top.  

If you want to overlay controls on TVideoPlayer in Kastri, you'll need to use the ["native" controls](https://github.com/DelphiWorlds/Kastri/tree/master/Demos/NativeControls), such as `TNativeImage`.

**Comment by firemen01** [comment]
Thank you very much for your answer!
some simple things turn out to be very complicated right away)
no matter how I try to upload a picture, there is only a blue square with text on the screen without an image, but one thing is good, it is now displayed on the video)
what am I doing wrong?

```
  Image:= TNativeImage.Create(myForm);
  Image.Text:= 'Tested';
  Image.Parent:= myForm;
  Image.BackgroundColor:= TAlphaColorRec.Blue;

  Image.Image.LoadFromFile(DocumentPath+'fire.png'); // not work

  var Bitmap: TBitmap;
  Bitmap:= TBitmap.Create;
  Bitmap.LoadFromFile(DocumentPath+'fire.png');
  
  Image.Image.Assign(Bitmap); // not work

  Image.Image.SetSize(Bitmap.Width, Bitmap.Height);
  Image.Image.CopyFromBitmap(Bitmap);  // not work
```

**Comment by DelphiWorlds** [comment]
> Image.Image.LoadFromFile(DocumentPath+'fire.png'); // not work

It should just be:

```delphi
Image.LoadFromFile(DocumentPath+'fire.png');
```

**Comment by firemen01** [comment]
it doesn't work either
https://file.io/l9HSgeEmqgge

**Comment by DelphiWorlds** [comment]
You need to turn on Secure file sharing:

<img width="459" alt="image" src="https://github.com/user-attachments/assets/03eab8b6-8740-4139-9bbd-c6c17b569369">

I'll update the readme to reflect this, thanks!

**Comment by firemen01** [comment]
thank you! it worked.
but still if I have images prepared in TBitmap how to assign?
it doesn't work:

  var Bitmap: TBitmap;
  Bitmap:= TBitmap.Create;
  Bitmap.LoadFromFile(DocumentPath+'fire.png');
  Image.Assign(Bitmap);

or

  Image.Image.SetSize(Bitmap.Width, Bitmap.Height);
  Image.Image.CopyFromBitmap(Bitmap);
or   
  Image.Image.Canvas.DrawBitmap(Bitmap, Rectf(0,0, Bitmap.Width, Bitmap.Height), Rectf(0,0, Bitmap.Width, Bitmap.Height), 1);

I would like to create my own bitmaps and draw them on the TNativeImage over the video
I think:
property Image: FMX.Graphics.TBitmap
but nothing works right away
in general, there are so many pitfalls with android even on simple things that while I'm using Alcione, everything works predictably there

**Comment by DelphiWorlds** [comment]
> Image.Assign(Bitmap);

I've now pushed changes that will fix this, however the code will need to be:

```delphi
Image.Image.Assign(Bitmap);
```

> while I'm using Alcione, everything works predictably there

Alcinoe has its fair share of issues.
